### PR TITLE
Merge CumsumOp/CumprodOp into CumOp

### PR DESCRIPTION
**Pull request by khaotik**
Changes:
 - Merged `CumsumOp` and `CumprodOp` into `CumOp`
 - Changed `GpuCumsum` into `GpuCumOp`, now `cumprod` is available on new GPU backend

Notes:
 - `cumsum` and `cumprod` convenience functions remain syntactically same
 - Should we change the name of this Op? (If you get it)

### CODE REVIEW

File: theano/tensor/extra_ops.py [hunk 1]
```diff
@@ -264,20 +265,37 @@ def make_node(self, x):
...
+        noimpl = NotImplementedError(
+            'CumOp: unknown gradient for mode %s' % self.mode)
 
         if self.axis is None:
-            return [cumsum(gi[::-1])[::-1].reshape(inputs[0].shape)]
+            if self.mode == 'add':
+                return [cumsum(gi[::-1])[::-1].reshape(x.shape)]
+            elif self.mode == 'mul':
+                return [cumsum(
+                    (fx * gi)[::-1])[::-1].reshape(x.shape) / x]
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: Move the creating of fx on the line above.

File: theano/tensor/extra_ops.py [hunk 2]
```diff
@@ -264,20 +265,37 @@ def make_node(self, x):
...
-        # We need to reverse the gradients along ``self.axis``,
-        #  compute cumsum, then reverse again
         reverse_slicing = [slice(None, None, None)] * gi.ndim
         reverse_slicing[self.axis] = slice(None, None, -1)
         reverse_slicing = tuple(reverse_slicing)
-        return [cumsum(gi[reverse_slicing], self.axis)[reverse_slicing]]
+        if self.mode == 'add':
+            return [cumsum(gi[reverse_slicing], self.axis)[reverse_slicing]]
+        elif self.mode == 'mul':
+            # We need to reverse the gradients along ``self.axis``,
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: Why did you move this comment?

File: theano/tensor/extra_ops.py [hunk 3]
```diff
@@ -264,20 +265,37 @@ def make_node(self, x):
...
+            #  compute cumsum, then reverse again
+            return [cumsum(
+                (fx * gi)[reverse_slicing], self.axis)[reverse_slicing] / x]
```
Comment (nouiz) [inline theano/tensor/extra_ops.py]: Can the grad of cumprod be writen as a cumprod? If so, this could make the graph simpler. I didn't look at the math.

Comment (khaotik) [inline theano/tensor/extra_ops.py]: The only improvement I can figure out is to introduce a "reverse" mode for CumOp:
```
cumop(vec, mode='add', reverse='True')
# same as reverse(cumop(reverse(vec), mode='add'))
```
This requires modifying `c_code`. The reverse slicing Ops can be removed this way. Still, there must be coupling cumsum and cumprod, and the buggy(#5197) division remains.

**Comment by nouiz** [comment]
I think all is good. Thank!
